File: backend/app/modules/marketplace/model/marketplace.py
import enum
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.infrastructure.db.models import (
    MarketplacePayoutStatus,
    MarketplaceSettlementStatus,
    PurchaseStatus,
    PromptPaymentMethod,
    ReviewModerationStatus,
)


class ReviewSort(str, enum.Enum):
    new = "new"
    best = "best"


class CatalogAction(str, enum.Enum):
    open = "open"
    buy = "buy"
    signin = "signin"


class PromptPriceRead(BaseModel):
    price_rub: int
    price_lumens: int
    commission_percent: int = 5


class PromptAccessRead(BaseModel):
    has_access: bool
    is_owned: bool = False
    source: str | None = None
    can_unlock_with_plan: bool = False
    remaining_plan_unlocks: int = 0
    monthly_plan_unlocks: int = 0
    purchase_required: bool = False
    catalog_action: CatalogAction = CatalogAction.open


class PromptReviewWrite(BaseModel):
    rating: int = Field(ge=1, le=5)
    text: str | None = Field(default=None, max_length=3000)


class PromptReviewRead(BaseModel):
    id: uuid.UUID
    rating: int
    text: str | None = None
    author_user_id: uuid.UUID
    author_display_name: str
    author_slug: str | None = None
    prompt_id: uuid.UUID
    prompt_slug: str
    prompt_title: str
    created_at: datetime
    updated_at: datetime
    verified_purchase: bool = True
    moderation_status: ReviewModerationStatus = ReviewModerationStatus.visible
    moderation_reason: str | None = None
    reported_count: int = 0


class PromptReviewListRead(BaseModel):
    seller_user_id: uuid.UUID | None = None
    rating_average: float | None = None
    rating_display: float | None = None
    review_count: int = 0
    sort: ReviewSort
    items: list[PromptReviewRead] = []


class PromptPurchaseRead(BaseModel):
    id: uuid.UUID
    prompt_id: uuid.UUID
    prompt_slug: str
    prompt_title: str
    seller_user_id: uuid.UUID | None = None
    status: PurchaseStatus
    payment_method: PromptPaymentMethod
    price_rub: int
    price_lumens: int
    settlement_status: MarketplaceSettlementStatus = MarketplaceSettlementStatus.pending
    settlement_available_at: datetime | None = None
    paid_out_at: datetime | None = None
    created_at: datetime
    completed_at: datetime | None = None
    can_review: bool = False
    review: PromptReviewRead | None = None


class PromptPurchaseListRead(BaseModel):
    items: list[PromptPurchaseRead] = []
    total: int = 0


class PromptPurchaseActionResponse(BaseModel):
    purchase: PromptPurchaseRead
    access: PromptAccessRead


class PromptCheckoutSessionRequest(BaseModel):
    prompt_id: uuid.UUID
    client_token: str | None = Field(default=None, min_length=8, max_length=80)
    success_url: str | None = Field(default=None, max_length=1000)
    cancel_url: str | None = Field(default=None, max_length=1000)


class PromptCheckoutSessionResponse(BaseModel):
    url: str
    session_id: str | None = None
    purchase_id: uuid.UUID


class PromptLumenPurchaseRequest(BaseModel):
    client_token: str | None = Field(default=None, min_length=8, max_length=80)


class TrustIndicatorRead(BaseModel):
    key: str
    level: Literal["info", "good", "strong"] = "info"


class MarketplacePayoutRead(BaseModel):
    id: uuid.UUID
    currency_code: str
    status: MarketplacePayoutStatus
    total_amount: int
    purchase_count: int
    external_reference: str | None = None
    requested_at: datetime
    paid_at: datetime | None = None


class MarketplacePayoutRequestWrite(BaseModel):
    currency_code: str = Field(min_length=3, max_length=8)
    notes: str | None = Field(default=None, max_length=500)

    @field_validator("currency_code")
    @classmethod
    def normalize_currency(cls, value: str) -> str:
        normalized = value.strip().upper()
        if normalized not in {"RUB", "LMN"}:
            raise ValueError("currency_code must be RUB or LMN")
        return normalized


class MarketplacePayoutFinalizeWrite(BaseModel):
    reference: str | None = Field(default=None, max_length=120)


class PromptReviewReportWrite(BaseModel):
    reason: str = Field(min_length=3, max_length=64)
    details: str | None = Field(default=None, max_length=500)


class SellerMarketplaceSummaryRead(BaseModel):
    rating_average: float | None = None
    rating_display: float | None = None
    review_count: int = 0
    sold_prompts_count: int = 0
    purchases_count: int = 0
    seller_revenue_rub: int = 0
    seller_lumens_earned: int = 0
    pending_balance_rub: int = 0
    available_balance_rub: int = 0
    paid_out_rub: int = 0
    refunded_balance_rub: int = 0
    disputed_balance_rub: int = 0
    pending_balance_lumens: int = 0
    available_balance_lumens: int = 0
    paid_out_lumens: int = 0
    refunded_balance_lumens: int = 0
    disputed_balance_lumens: int = 0
    platform_commission_rub: int = 0
    platform_commission_lumens: int = 0
    clawback_due_rub: int = 0
    clawback_due_lumens: int = 0
    payout_eligible: bool = False
    trust_indicators: list[TrustIndicatorRead] = []
    recent_reviews: list[PromptReviewRead] = []
    recent_payouts: list[MarketplacePayoutRead] = []


class MarketplaceOverviewRead(BaseModel):
    summary: SellerMarketplaceSummaryRead
    purchases: list[PromptPurchaseRead] = []
    reviews: list[PromptReviewRead] = []
    payouts: list[MarketplacePayoutRead] = []

File: frontend/lib/types/marketplace.ts
import type { PurchaseStatus } from "./economy";
import type { TrustIndicator } from "./shared";

export type ReviewSort = "new" | "best";
export type ReviewModerationStatus = "visible" | "pending" | "hidden";
export type MarketplaceSettlementStatus = "pending" | "available" | "paid_out" | "refunded" | "disputed";
export type MarketplacePayoutStatus = "requested" | "processing" | "paid" | "failed" | "canceled";

export type PromptReview = {
  id: string;
  rating: number;
  text: string | null;
  author_user_id: string;
  author_display_name: string;
  author_slug: string | null;
  prompt_id: string;
  prompt_slug: string;
  prompt_title: string;
  created_at: string;
  updated_at: string;
  verified_purchase: boolean;
  moderation_status?: ReviewModerationStatus;
  moderation_reason?: string | null;
  reported_count?: number;
};

export type PromptReviewList = {
  seller_user_id?: string | null;
  rating_average?: number | null;
  rating_display?: number | null;
  review_count: number;
  sort: ReviewSort;
  items: PromptReview[];
};

export type PromptMarketplacePurchase = {
  id: string;
  prompt_id: string;
  prompt_slug: string;
  prompt_title: string;
  seller_user_id: string | null;
  status: PurchaseStatus;
  payment_method: "included_limit" | "lumens" | "stripe" | "legacy_store";
  price_rub: number;
  price_lumens: number;
  settlement_status?: MarketplaceSettlementStatus;
  settlement_available_at?: string | null;
  paid_out_at?: string | null;
  created_at: string;
  completed_at: string | null;
  can_review: boolean;
  review?: PromptReview | null;
};

export type MarketplacePayout = {
  id: string;
  currency_code: string;
  status: MarketplacePayoutStatus;
  total_amount: number;
  purchase_count: number;
  external_reference?: string | null;
  requested_at: string;
  paid_at?: string | null;
};

export type SellerMarketplaceSummary = {
  rating_average: number | null;
  rating_display: number | null;
  review_count: number;
  sold_prompts_count: number;
  purchases_count: number;
  seller_revenue_rub: number;
  seller_lumens_earned: number;
  pending_balance_rub: number;
  available_balance_rub: number;
  paid_out_rub: number;
  refunded_balance_rub: number;
  disputed_balance_rub: number;
  pending_balance_lumens: number;
  available_balance_lumens: number;
  paid_out_lumens: number;
  refunded_balance_lumens: number;
  disputed_balance_lumens: number;
  platform_commission_rub: number;
  platform_commission_lumens: number;
  clawback_due_rub: number;
  clawback_due_lumens: number;
  payout_eligible: boolean;
  trust_indicators: TrustIndicator[];
  recent_reviews: PromptReview[];
  recent_payouts: MarketplacePayout[];
};

export type MarketplaceOverview = {
  summary: SellerMarketplaceSummary;
  purchases: PromptMarketplacePurchase[];
  reviews: PromptReview[];
  payouts: MarketplacePayout[];
};
